src/Dependencies/PooledObjects/SpannableArrayBuilder.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.PooledObjects;
+
+[DebuggerDisplay("Count = {Count,nq}")]
+internal sealed class SpannableArrayBuilder<T>
+{
+    public const int PooledArrayLengthLimitExclusive = 128;
+
+    private static readonly ObjectPool<SpannableArrayBuilder<T>> s_pool = new(() => new(capacity: 4), size: 16);
+
+    private T[] _items;
+    private int _count;
+
+    internal SpannableArrayBuilder()
+        : this(capacity: 8)
+    {
+    }
+
+    internal SpannableArrayBuilder(int capacity)
+    {
+        _items = new T[capacity];
+        _count = 0;
+    }
+
+    public static SpannableArrayBuilder<T> GetInstance()
+        => s_pool.Allocate();
+
+    public void Free()
+    {
+        if (_items.Length < PooledArrayLengthLimitExclusive)
+        {
+            if (Count != 0)
+                Clear();
+
+            s_pool.Free(this);
+        }
+        else
+        {
+            s_pool.ForgetTrackedObject(this);
+        }
+    }
+
+    public void Add(T item)
+    {
+        var newCount = _count + 1;
+        EnsureCapacity(newCount);
+        _items[_count] = item;
+        _count = newCount;
+    }
+
+    public void AddRange(ReadOnlySpan<T> items)
```

Comment:
nit: could be an extension method.

Reply:
Going to let the compiler folks chime in whether they would prefer that. This follows ArrayBuilder pretty closely, and that class defines it directly.